### PR TITLE
Add Elastic-Stack chart

### PR DESCRIPTION
**Pull request by rendhalver**
This is dependent on a couple of other PR's.
DNS discovery for Elasticsearch #2889
nginx-ldapauth-proxy #3013
logstash fixes here: #3061
And these PR's on the upstream curator chart: samsung-cnct/chart-curator#11 and  samsung-cnct/chart-curator#13

*****************************************************************************************

Thank you for contributing to kubernetes/charts. Before you submit this PR we'd like to
make sure you are aware of our technical requirements and best practices:

* https://github.com/kubernetes/charts/blob/master/CONTRIBUTING.md#technical-requirements
* https://github.com/kubernetes/helm/tree/master/docs/chart_best_practices

For a quick overview across what we will look at reviewing your PR, please read
our review guidelines:

* https://github.com/kubernetes/charts/blob/master/REVIEW_GUIDELINES.md

Following our best practices right from the start will accelerate the review process and
help get your PR merged quicker.

When updates to your PR are requested, please add new commits and do not squash the 
history. This will make it easier to identify new changes. The PR will be squashed 
anyways when it is merged. Thanks.

*****************************************************************************************

### CODE REVIEW

File: incubator/elk/requirements.yaml [hunk 1]
```diff
@@ -0,0 +1,16 @@
+dependencies:
+- name: elasticsearch
+  version: 0.5.0
+  repository: https://kubernetes-charts-incubator.storage.googleapis.com/
+- name: kibana
+  version: 0.2.0
+  repository: https://kubernetes-charts.storage.googleapis.com/
+- name: logstash
+  version: 0.3.1
```
Comment (christian-roggia) [inline incubator/elk/requirements.yaml]: See [Requirements Files](https://github.com/kubernetes/helm/blob/648239edc3d2ee0adb607cecb859b013b7f3d86e/docs/chart_best_practices/requirements.md)
```yaml
version: ^0.3.1
```

File: incubator/elk/requirements.yaml [hunk 2]
```diff
@@ -0,0 +1,16 @@
+dependencies:
+- name: elasticsearch
+  version: 0.5.0
+  repository: https://kubernetes-charts-incubator.storage.googleapis.com/
+- name: kibana
+  version: 0.2.0
```
Comment (christian-roggia) [inline incubator/elk/requirements.yaml]: See [Requirements Files](https://github.com/kubernetes/helm/blob/648239edc3d2ee0adb607cecb859b013b7f3d86e/docs/chart_best_practices/requirements.md)
```yaml
version: ^0.2.0
```

File: incubator/elk/requirements.yaml [hunk 3]
```diff
@@ -0,0 +1,16 @@
+dependencies:
+- name: elasticsearch
+  version: 0.5.0
```
Comment (christian-roggia) [inline incubator/elk/requirements.yaml]: See [Requirements Files](https://github.com/kubernetes/helm/blob/648239edc3d2ee0adb607cecb859b013b7f3d86e/docs/chart_best_practices/requirements.md)
```yaml
version: ^0.5.0
```

File: incubator/elk/requirements.yaml [hunk 4]
```diff
@@ -0,0 +1,16 @@
...
+  repository: https://kubernetes-charts-incubator.storage.googleapis.com/
+- name: kibana
+  version: 0.2.0
+  repository: https://kubernetes-charts.storage.googleapis.com/
+- name: logstash
+  version: 0.3.1
+  repository: https://kubernetes-charts-incubator.storage.googleapis.com/
+- name: nginx-ldapauth-proxy
+  version: 0.1.1
```
Comment (christian-roggia) [inline incubator/elk/requirements.yaml]: See [Requirements Files](https://github.com/kubernetes/helm/blob/648239edc3d2ee0adb607cecb859b013b7f3d86e/docs/chart_best_practices/requirements.md)
```yaml
version: ^0.1.1
```

File: incubator/elk/requirements.yaml [hunk 5]
```diff
@@ -0,0 +1,16 @@
...
+  repository: https://kubernetes-charts-incubator.storage.googleapis.com/
+- name: curator
+  version: 0.1.0
```
Comment (christian-roggia) [inline incubator/elk/requirements.yaml]: See [Requirements Files](https://github.com/kubernetes/helm/blob/648239edc3d2ee0adb607cecb859b013b7f3d86e/docs/chart_best_practices/requirements.md)
```yaml
version: ^0.1.0
```

File: incubator/elk/Chart.yaml [hunk 6]
```diff
@@ -0,0 +1,9 @@
+apiVersion: v1
+description: A Helm chart for ELK
```
Comment (christian-roggia) [inline incubator/elk/Chart.yaml]: From [The Open Source Elastic Stack](https://www.elastic.co/products):
```yaml
description: Reliably and securely take data from any source, in any format, and 
search, analyze, and visualize it in real time.
home: https://www.elastic.co/products
```

File: incubator/elk/Chart.yaml [hunk 7]
```diff
@@ -0,0 +1,9 @@
...
+name: elk
```
Comment (christian-roggia) [inline incubator/elk/Chart.yaml]: I'd probably change the name to either:
```yaml
name: elk-stack
```
or 
```yaml
name: elastic-stack
```
See [An Introduction to the ELK Stack (Now the Elastic Stack)](https://www.elastic.co/webinars/introduction-elk-stack)

File: incubator/elk/requirements.lock [hunk 8]
```diff
@@ -0,0 +1,9 @@
+dependencies:
+- name: kibana
+  repository: https://kubernetes-charts.storage.googleapis.com/
+  version: 0.2.0
+- name: logstash
+  repository: https://kubernetes-charts-incubator.storage.googleapis.com/
+  version: 0.3.0
```
Comment (christian-roggia) [inline incubator/elk/requirements.lock]: Version does not match the `requirements.yaml` version. Please update the dependencies with the helm command `helm dependency update`.

**Comment by christian-roggia** [comment]
To solve the CI error it should be enough to add an empty `templates/NOTES.txt` file to the chart (therefore you can remove the `templates/.keep` file). Please consider to add useful information there as soon as the chart is ready.

Also the README file is currently missing, I can provide one if you are busy setting up the chart.

**Review by christian-roggia — CHANGES_REQUESTED** [review]

**Comment by rendhalver** [comment]
Thanks @christian-roggia I will get that updated.

File: incubator/elk/requirements.lock (hunk 8 above)
Comment (rendhalver) [reply incubator/elk/requirements.lock]: Damn. I will fix.

**Comment by timstoop** [comment]
While you're at it, could you also implement the Java options that allow the JVM to check the cgroup settings for determining memory usage? More info here: https://blogs.oracle.com/java-platform-group/java-se-support-for-docker-cpu-and-memory-limits

(I came across this in PR 5418 and I think this is a really nice fix for almost all Java based applications.)

**Comment by rendhalver** [comment]
> While you're at it, could you also implement the Java options that allow the JVM to check the cgroup > settings for determining memory usage? More info here: https://blogs.oracle.com/java-platform-group/java-se-support-for-docker-cpu-and-memory-limits
>
> (I came across this in PR 5418 and I think this is a really nice fix for almost all Java based applications.)

@timstoop 
We should implement those changes in PR's on the sub-charts used by this Chart.

**Comment by rendhalver** [comment]
Can I get some help working out why the build failed?

**Comment by unguiculus** [comment]
A look at the CircleCI log reveals that the home field is missing in Chart.yml.

**Comment by rendhalver** [comment]
Ahh. Thanks @unguiculus 
I had a look for other charts using home and failed to find any. I guess my search terms wern't right.
I will get that added.

**Comment by rendhalver** [comment]
This has been waiting for a status from ci/circleci for a while. What test sends that status?

**Comment by rendhalver** [comment]
@christian-roggia I believe I have addressed your concerns. Do you have any others?

**Comment by christian-roggia** [comment]
This PR looks good to me but I would like to wait for the Elasticsearch DNS discovery PRs to be merged #5663 #2889. The ideal case would be to wait for the ES/ES-curator charts to be stable, but I don't expect that to happen anytime soon and incubator is for unstable charts anyway. I would also like to hear from the official ES dev team as their discussion about Kubernetes and Helm should be over.

**Comment by christian-roggia** [comment]
A suggestion on my side is to add the OWNERS file and to use PNG/JPEG images instead of SVGs (they are not displayed on kubeapps for some reasons).

**Comment by rendhalver** [comment]
Good point. I will add an OWNERS file and find better images.
If we leave this in incubator we can just add another PR when DNS discovery gets merged.
Once ES/ES-curator are stable we can promote this.

**Comment by rendhalver** [comment]
Damn. Elastic don't see to use non SVGs of the icons. :(

**Comment by timstoop** [comment]
@christian-roggia Just wondering why you want to other PRs to be merged first? Are you expecting them to be a hassle to merge with this PR?

**Comment by christian-roggia** [comment]
@timstoop This chart has dependencies on 4 different **incubator** charts, which make this chart already extremely unstable. I personally would like to wait for the PR on the DNS discovery to be merged so that this chart does not depend on open PRs (and therefore on known imminent breaking changes) too.

On the other hand, as I already told in my previous comment, this repository is for unstable charts. If you want to go ahead and merge this PR please do so, I just wanted to express a few words of concern about the extreme instability of this chart.

**Comment by rendhalver** [comment]
This chart is still in in incubator itself.
I only mentioned the DNS discovery PR because it's a more stable way to deploy ES.
The chart itself only depends on the released versions of it's sub-charts.

**Comment by rendhalver** [comment]
We are also using this in production to deploy all of our ES clusters using the DNS discovery and it works well.
Maybe that's a case for getting the DNS discovery chart merged first.

**Comment by rendhalver** [comment]
@christian-roggia and @timstoop I updated the dependencies to the latest which includes the DNS discovery for Elasticsearch merged in #5663 and we have passing tests.
Are there any further concerns to getting this merged soon?

**Comment by christian-roggia** [comment]
It looks good to me.

File: incubator/elastic-stack/templates/_helpers.tpl [hunk 9]
```diff
@@ -0,0 +1,16 @@
+{{/* vim: set filetype=mustache: */}}
```
Comment (unguiculus) [inline incubator/elastic-stack/templates/_helpers.tpl:1]: Please update with latest `_helpers.tpl` you get from `helm create`.

File: incubator/elastic-stack/values.yaml [hunk 10]
```diff
@@ -0,0 +1,19 @@
+# Default values for elk.
+# This is a YAML-formatted file.
+# Declare variables to be passed into your templates.
+
+kibana:
+  env:
+    ELASTICSEARCH_URL: http://http.default.svc.cluster.local:9200
+
+logstash:
```
Comment (unguiculus) [inline incubator/elastic-stack/values.yaml:9]: What's the logic behind these defaults?

**Review by unguiculus — CHANGES_REQUESTED** [review]
Please add readme and some contents to NOTES.txt.

File: incubator/elastic-stack/values.yaml (hunk 10 above)
Comment (rendhalver) [reply incubator/elastic-stack/values.yaml:9]: These are just the defaults for the sub-charts.

**Comment by rendhalver** [comment]
@christian-roggia Can I take you up on your offer of a making a README?

**Comment by christian-roggia** [comment]
This week I will be busy but I can work on it next week.

**Comment by rendhalver** [comment]
I have some questions about what should go in the NOTES file for this.
Is there a way to display the notes from each of the dependencies?

File: incubator/elastic-stack/templates/_helpers.tpl (hunk 9 above)
Comment (rendhalver) [reply incubator/elastic-stack/templates/_helpers.tpl:1]: Done.

**Comment by rendhalver** [comment]
@unguiculus I have added basic NOTES and README files.

File: incubator/elastic-stack/README.md [hunk 11]
```diff
@@ -0,0 +1,44 @@
+# Elastic-stack Helm Chart
+
+This chart installs an elasticsearch cluster with kibana and logstash by default.
+You can optionally disable logstash and install Fluentd if you prefer. It also optionally installs nginx-ldapauth-proxy and elasticsearch-curator.
+
+## Prerequisites Details
+
+* Kubernetes 1.6+
```
Comment (unguiculus) [inline incubator/elastic-stack/README.md]: 1.6? Really? I didn't check the dependent chart, but generally, we support the current and previous minor versions of Kubernetes.

**Review by unguiculus — CHANGES_REQUESTED** [review]

File: incubator/elastic-stack/README.md (hunk 11 above)
Comment (rendhalver) [reply incubator/elastic-stack/README.md]: Ah true.
I copied that from the elasticsearch chart which may have been written when 1.8 was the supported release.
I will update it.

**Comment by rendhalver** [comment]
@unguiculus and @christian-roggia Do either of you have any further concerns?

**Review by unguiculus — APPROVED** [review]

**Pull request merged**